numpyro/infer/util.py

```diff
@@ -660,7 +660,7 @@ def initialize_model(
         data={
             k: site["value"]
             for k, site in model_trace.items()
-            if site["type"] in ["param"]
+            if site["type"] in ["param", "mutable"]
```

Comment:
maybe also prng_key, so that random keys can be used for dropout

Reply:
As in `if site["type"] in ["param", "mutable", "prng_key"]` If I do this and ever the dropout changes `numpyro.prng_key()` - > `random.PRNGKey(0)` I get the error

```python
def apply_without_state(params, *args, **kwargs):
>       return nn_module.apply({"params": params}, *args, **kwargs)
E       ValueError: The ``rngs`` argument passed to an apply function should be a ``jax.PRNGKey`` or a dictionary mapping strings to ``jax.PRNGKey``.
E       --------------------
```

Or am I missing something 😅 ?

Reply:
maybe we need to allow substitute rng_key: https://github.com/pyro-ppl/numpyro/blob/5af9ebda72bd7aeb08c61e4248ecd0d982473224/numpyro/handlers.py#L808

Reply:
If I add it as 

```python
msg["type"] not in ("sample", "param", "mutable", "plate", "deterministic", "prng_key")
```

The tests fail with 

```python
    def process_message(self, msg):
        if (
            msg["type"] not in ("sample", "param", "mutable", "plate", "deterministic", "prng_key")
        ) or msg.get("_control_flow_done", False):
            if msg["type"] == "control_flow":
                if self.data is not None:
                    msg["kwargs"]["substitute_stack"].append(("substitute", self.data))
                if self.substitute_fn is not None:
                    msg["kwargs"]["substitute_stack"].append(
                        ("substitute", self.substitute_fn)
                    )
            return
    
        if self.data is not None:
>           value = self.data.get(msg["name"])
E           KeyError: 'name'
```

Reply:
Or I I remove "mutable" I get the same error

```python
    def apply_without_state(params, *args, **kwargs):
>       return nn_module.apply({"params": params}, *args, **kwargs)
E       ValueError: The ``rngs`` argument passed to an apply function should be a ``jax.PRNGKey`` or a dictionary mapping strings to ``jax.PRNGKey``.
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.
```

Reply:
oh, prng_key has no name :( let me think a bit more how to define a static model with PRNGKey in the initialize_model utility.

Reply:
Sure! No rush! Is now late in Berlin anyway 😴😄

Reply:
Maybe using `substitute_fn` instead of `data` in initialize_model, where we return random.PRNGKey(0) if msg["type"] == "prng_key"

Reply:
I tried trying what (I think) you suggested in https://github.com/pyro-ppl/numpyro/pull/1823/commits/a37f7cea9d252ea2cc362a3a9346ab7f90de597a But I keep getting the `KeyError: 'name'`. Please let me know if I missed something or have a tip on how to fix this one :)